From 8361dd2f2100d64aa9500268865addafc5a42785 Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Mon, 17 Apr 2023 03:27:36 -0700
Subject: [PATCH 56/76] UBUNTU: SAUCE: apparmor4.0.0 [56/76]: cleanup: provide
 separate audit messages for file and policy checks

Improve policy load failure messages by identifying which dfa the
verification check failed in.

Signed-off-by: John Johansen <john.johansen@canonical.com>
---
 security/apparmor/policy_unpack.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/security/apparmor/policy_unpack.c b/security/apparmor/policy_unpack.c
index 690ec6b2d0bd..9fdb9cc15e3e 100644
--- a/security/apparmor/policy_unpack.c
+++ b/security/apparmor/policy_unpack.c
@@ -1292,12 +1292,18 @@ static int verify_profile(struct aa_profile *profile)
 	if (!rules)
 		return 0;
 
-	if ((rules->file.dfa && !verify_dfa_accept_index(rules->file.dfa,
-							 rules->file.size)) ||
-	    (rules->policy.dfa &&
-	     !verify_dfa_accept_index(rules->policy.dfa, rules->policy.size))) {
+	if (rules->file.dfa && !verify_dfa_accept_index(rules->file.dfa,
+							rules->file.size)) {
 		audit_iface(profile, NULL, NULL,
-			    "Unpack: Invalid named transition", NULL, -EPROTO);
+			    "Unpack: file Invalid named transition", NULL,
+			    -EPROTO);
+		return -EPROTO;
+	}
+	if (rules->policy.dfa &&
+	    !verify_dfa_accept_index(rules->policy.dfa, rules->policy.size)) {
+		audit_iface(profile, NULL, NULL,
+			    "Unpack: policy Invalid named transition", NULL,
+			    -EPROTO);
 		return -EPROTO;
 	}
 
-- 
2.42.0

